internal/adminapi/tls.go

```diff
@@ -0,0 +1,71 @@
+package adminapi
+
+import (
+	"crypto/tls"
+	"fmt"
+	"os"
+)
+
+// TLSClientConfig contains TLS client certificate and client key to be used when connecting with Admin APIs.
+// It's validated with manager.validateClientTLS before passing it further down. It guarantees that only the
+// allowed combinations of variables will be passed:
+// - only one of Cert / CertFile
+// - only one of Key / KeyFile
+// - if any of Cert / CertFile is set, one of Key / KeyFile has to be set
+// - if any of Key / KeyFile is set, one of Cert / CertFile has to be set
+type TLSClientConfig struct {
+	// Cert is a client certificate.
+	Cert string
+	// CertFile is a client certificate file path.
+	CertFile string
+
+	// Key is a client key.
+	Key string
+	// KeyFile is a client key file path.
+	KeyFile string
+}
+
+func (c TLSClientConfig) IsZero() bool {
+	return c == TLSClientConfig{}
+}
+
+// extractClientCertificates extracts tls.Certificates from TLSClientConfig.
+// It returns an empty slice in case there was no client cert and/or client key provided.
+func extractClientCertificates(tlsClient TLSClientConfig) ([]tls.Certificate, error) {
+	clientCert, err := valueFromVariableOrFile(tlsClient.Cert, tlsClient.CertFile)
+	if err != nil {
+		return nil, fmt.Errorf("could not extract TLS client cert")
+	}
+	clientKey, err := valueFromVariableOrFile(tlsClient.Key, tlsClient.KeyFile)
+	if err != nil {
+		return nil, fmt.Errorf("could not extract TLS client key")
+	}
+
+	if len(clientCert) != 0 && len(clientKey) != 0 {
+		cert, err := tls.X509KeyPair(clientCert, clientKey)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load client certificate: %w", err)
+		}
+		return []tls.Certificate{cert}, nil
+	}
+
+	return nil, nil
+}
+
+// valueFromVariableOrFile uses v value if it's not empty, and falls back to reading a file content when value is missing.
+// When both are empty, nil is returned.
+func valueFromVariableOrFile(v string, file string) ([]byte, error) {
```

Comment:
Since we need ot convert this to `[]byte` anyway, how about using that type instead of `string` which can easily be mistaken e.g. with the `file` arg here.

Reply:
Similarly, I'd consider changing the types in `TLSClientConfig` accordingly. WDYT?

Reply:
> Since we need ot convert this to []byte anyway, how about using that type instead of string which can easily be mistaken e.g. with the file arg here.

Yup, totally agree. Changed that to `[]byte`.

> Similarly, I'd consider changing the types in TLSClientConfig accordingly. WDYT?

It won't be that simple to replace them since `pflag` doesn't have a built-in variable type for flags that should be directly converted from `string` -> `[]byte`, there are only `BytesHex` and `BytesBase64` which do not satisfy our needs.